lale/grammar.py

```diff
@@ -0,0 +1,115 @@
+from lale.operators import MetaModelOperator, PlannedOperator, Operator, BasePipeline, OperatorChoice, IndividualOp
+from lale.operators import make_choice, make_pipeline
+from lale.lib.lale import NoOp
+from typing import Optional
+
+class Primitive(Operator):
+    """ Abstract operator for non-terminal grammar rules.
+    """
+    def __init__(self, name, value=None):
+        self._name = name
+        self.value = value
+        
+    def _lale_clone(self):
+        pass
+    
+    def has_same_impl(self):
+        pass
+    
+    def is_supervised(self):
+        return False
+
+    def name(self):
+        return self._name
+    
+    def set_name(self, name):
+        self._name = name
+        
+        
+class Grammar(MetaModelOperator):
+    """ Base class for Lale grammars. The for every grammar `g`, rule `g.start` must be defined.
+    """
+    def __init__(self):
+        self._primitives = {}
+
+    def __getattr__(self, name):
+        if name.startswith('_'):
+            return self.__dict__[name]
+        if name not in self._primitives:
+            self._primitives[name] = Primitive(name)
+        return self._primitives[name]
+        
+    def __setattr__(self, name, value):
+        if name.startswith('_'):
+            self.__dict__[name] = value
+        else:
+            self._primitives[name] = value
+            
+    def _lale_clone(self):
+        pass
+    
+    def has_same_impl(self):
+        pass
+    
+    def is_supervised(self):
+        return False
+
+    def name(self):
+        return self._name
+    
+    def set_name(self, name):
+        self._name = name
+            
+    def auto_arrange(self, planner):
+        pass
+    
+    def arrange(self, *args, **kwargs):
+        pass
+            
+            
+def unroll(g: Grammar, op: Primitive, n: int) -> Optional[PlannedOperator]:
+    """ Unroll all possible operators from the grammar `g` starting from non-terminal `op` after `n` derivations.
+    
+    Parameters
+    ----------
+    g : Grammar
+        input grammar
+    op : Primitive
+        starting rule (e.g., `g.start`)
+    n : int
+        number of derivations
+    
+    Returns
+    -------
+    Optional[PlannedOperator]
+    """
+    if isinstance(op, BasePipeline):
+        steps = [unroll(g, sop, n) for sop in op._steps]
+        return make_pipeline(*steps) if not None in steps else None
```

Comment:
What about the edges? I think as written, those get lost. I think this code turns all pipelines into linear sequences, which would be wrong for other graph shapes such as those created with the `&` combinator.

Reply:
Should be fixed in the next iteration.